File: 16-3sum-closest/3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int ans = nums[0] + nums[1] + nums[2];

        for (int i=0; i<nums.size(); i+=1){
            int l = i+1, r = nums.size()-1;

            while (l < r){
                int sum = nums[i] + nums[l] + nums[r];

                if (sum == target) return target;
                else if (abs(sum-target) < abs(target - ans)) {
                    ans = sum;
                }
                else if (sum < target) l+=1;
                else if (sum > target) r-=1;
            }
        }

        return ans;
    }
};

File: 16-3sum-closest/3sum-closest.go
func threeSumClosest(nums []int, target int) int {
    sort.Ints(nums)
    closest := 0
    min_diff := math.MaxInt

    for i := range nums {
        if i != 0 && nums[i] == nums[i-1] {
            continue
        }
        l, r := i+1, len(nums)-1
        for l < r {
            sum := nums[i] + nums[l] + nums[r]

            if sum == target {
                return sum
            }

            diff := abs(target - sum)
            // Update the closest
            if diff < min_diff {
                closest = sum
                min_diff = diff
            }
            // Update the pointers
            if sum > target {
                r--
            } else {
                l++
            }
        }
    }

    return closest
}
func abs(x int) int {
    if x < 0 {
        return -1 * x
    }
    return x
}